<launch>
    <arg name="object_avoidance" default="false"/>
    <arg name="endless" default="false"/>
    <arg name="waypoint_file" default=""/>

    <include file="$(find kcctcore)/launch/include/costmap.launch"/>
    <include file="$(find waypoint_tools)/launch/path_display.launch">
        <arg name="wayFile" value="$(arg waypoint_file)"/>
    </include>

    <node name="wp_select" pkg="waypoint_tools" type="wp_select">
        <param name="base_link_frame_id" value="base_link"/>
        <param name="map_frame_id" value="map"/>
        <param name="loop_rate" value="20"/>
        <param name="target_deviation" value="2.5"/>
        <param name="precision_target_deviation" value="0.8"/>
        <param name="final_target_deviation" value="0.4"/>
        <param name="endless" value="$(arg endless)"/>
        <remap from="path" to="waypoint/path"/>
        <remap from="successPlan" to="astar_planning_node/successPlan"/>
    </node>

   <node name="astar_planning_node" pkg="kcctnavigation" type="astar_planning_node" respawn="true" output="screen">
        <rosparam file="$(find kcctcore)/config/nav_param/astar_planning.yaml" command="load" />
        <remap from="astar_plannnig_node/goal" to="nowWpPose"/>
        <remap from="astar_plannnig_node/costmap" to="costmap_node/marged_costmap"/><!--hitoyke; marged_costmap-->
    </node>

        <node name="lane_planner" pkg="kcctnavigation" type="lane_planner" >
            <rosparam file="$(find kcctcore)/config/nav_param/lane_planner.yaml" command="load" />
            <remap from="lane_planner/costmap" to="costmap_node/my_costmap/costmap"/>
            <remap from="path" to="astar_plannnig_node/path"/>
            <remap from="odom" to="/odom"/>
        </node>

        <node name="wpBack_recovery" pkg="kcctnavigation" type="wpBack_recovery">
            <param name="base_link_frame_id" value="base_link"/>
            <param name="map_frame_id" value="map"/>
            <param name="loop_rate" value="20"/>
            <param name="linear_vel" value="-0.3"/>
            <param name="max_angular_vel" value="0.5"/>
            <param name="duration" value="3.0"/>
            <param name="recovery_leastDistance" value="1.0"/>
            <param name="fin_recovery_deviation" value="0.2"/>
            <remap from="path" to="waypoint/path"/>
            <remap from="wpBack_recovery/mode" to="wpBack_recovery/mode"/>
        </node>
    </group>

    <!--物体回避無効-->
    <group unless="$(arg object_avoidance)">
        <node name="astar_planning_node" pkg="kcctnavigation" type="astar_planning_node" respawn="true">
            <rosparam file="$(find kcctcore)/config/nav_param/astar_planning.yaml" command="load" />
            <remap from="astar_plannnig_node/goal" to="nowWpPose"/>
            <remap from="astar_plannnig_node/costmap" to="costmap_node/path_costmap"/>
        </node>

        <node name="lane_planner" pkg="kcctnavigation" type="lane_planner" >
            <rosparam file="$(find kcctcore)/config/nav_param/lane_planner.yaml" command="load" />
            <remap from="lane_planner/costmap" to="costmap_node/path_costmap"/>
            <remap from="path" to="astar_plannnig_node/path"/>
            <remap from="odom" to="/odom"/>
        </node>
        
        <node name="wpBack_recovery" pkg="kcctnavigation" type="wpBack_recovery">
            <param name="base_link_frame_id" value="base_link"/>
            <param name="map_frame_id" value="map"/>
            <param name="loop_rate" value="20"/>
            <param name="linear_vel" value="-0.3"/>
            <param name="max_angular_vel" value="0.5"/>
            <param name="duration" value="3.0"/>
            <param name="recovery_leastDistance" value="1.0"/>
            <param name="fin_recovery_deviation" value="0.2"/>
            <remap from="path" to="waypoint/path"/>
            <remap from="wpBack_recovery/mode" to="wpBack_recovery/mode_unuse"/>
        </node>
    </group>

    <node name="pure_posePursuit" pkg="kcctnavigation" type="pure_posePursuit">
        <param name="base_link_frame_id" value="base_link"/>
        <param name="map_frame_id" value="map"/>
        <param name="loop_rate" value="30"/>
        <param name="maxVelocity" value="0.6"/><!--speed-->
        <param name="max_angular_vel" value="0.5"/>
        <param name="maxCurvature" value="3"/>
        <remap from="pure_posePursuit/pose_in" to="lane_planner/pose_out"/>
        <remap from="pure_posePursuit/cmd_vel" to="mcl_cmd_vel"/>
    </node>

    <!--safety function-->
    <node pkg="kcctnavigation" name="safety_limit" type="safety_limit" output="screen">
        <rosparam file="$(find kcctcore)/config/nav_param/safety_limit.yaml" command="load" />
        <remap from="safety_limit/cmd_vel_in" to="selected_cmd_vel"/>
        <remap from="safety_limit/cmd_vel_out" to="final_cmd_vel"/>
        <remap from="cloud" to="laser2pc/pc2"/>
        <remap from="safety_limit/costmap" to="costmap_node/my_costmap/costmap"/>
     </node>

    <!--MASTER NODE-->
    <node name="navigation_master" pkg="kcctcore" type="navigation_master" respawn="true" output="screen">
        <param name="base_link_frame_id" value="base_link"/>
        <param name="map_frame_id" value="map"/>
        <param name="loop_rate" value="30"/>
        <remap from="twist_maneger/targetWpPose_in" to="nowWpPose"/>
        <remap from="twist_maneger/targetPose_in" to="lane_planner/pose_out"/>
        <remap from="recovery/cmd_vel" to="wpBack_recovery/cmd_vel"/>
        <remap from="recovery/mode" to="wpBack_recovery/mode"/>
    </node>
</launch>